### DISCUSSION TITLE
Logging configuration override could use improvements

### DISCUSSION THREAD

**Discussion by chbndrhnns** [br3ndonland/inboard#27] [Q&A] [answered]
I noticed that overriding the logging configuration fails in two cases. Basically, I have added an `app` package in a [branch](https://github.com/br3ndonland/inboard/compare/develop...chbndrhnns:pythonpath) which hosts my FastAPI application. 

I am running the Docker image like this:

`docker run -it -v $(pwd)/app:/app/app --rm -e "LOGGING_CONF=app/logging_conf.py" inboard-logging`

a) specifying the configuration module with the dotted-path notation (like `LOGGING_CONF=app.core.logging_conf`) gives me this error:

```
Traceback (most recent call last):
  File "/app/inboard/start.py", line 127, in <module>
    logging_conf_dict = configure_logging(logger=logger)
  File "/app/inboard/start.py", line 41, in configure_logging
    if hasattr(logging_conf_module, "LOGGING_CONFIG"):
UnboundLocalError: local variable 'logging_conf_module' referenced before assignment
```

b) Duplicating `logging_conf.py` into my app which uses a custom formatter class fails due to discovery issues. It can be fixed by calling `start.py` with the dot notation instead of the path notation:

```
(.venv) ➜  inboard git:(pythonpath) docker run -it -v $(pwd)/app:/app/app --rm -e "LOGGING_CONF=app/logging_conf.py"  inboard-logging /bin/bash

root@bfc63406f754:/app# python inboard/start.py
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/logging/config.py", line 93, in _resolve
    found = getattr(found, n)
AttributeError: module 'app' has no attribute 'my_logger'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/logging/config.py", line 543, in configure
    formatters[name] = self.configure_formatter(
  File "/usr/local/lib/python3.9/logging/config.py", line 675, in configure_formatter
    c = _resolve(cname)
  File "/usr/local/lib/python3.9/logging/config.py", line 95, in _resolve
    __import__(used)
ModuleNotFoundError: No module named 'app.my_logger'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/app/inboard/start.py", line 127, in <module>
    logging_conf_dict = configure_logging(logger=logger)
  File "/app/inboard/start.py", line 46, in configure_logging
    logging.config.dictConfig(logging_conf_dict)
  File "/usr/local/lib/python3.9/logging/config.py", line 809, in dictConfig
    dictConfigClass(config).configure()
  File "/usr/local/lib/python3.9/logging/config.py", line 546, in configure
    raise ValueError('Unable to configure '
ValueError: Unable to configure formatter 'simple'

root@bfc63406f754:/app# python -m inboard.start
[prestart] Hello World, from prestart.py! Add database migrations and other scripts here.
[...]
KeyboardInterrupt
```

**Comment by br3ndonland** [comment, marked as answer]
Thank you for your report. A few suggestions that may help:

- In your first example, the module path is set to `app.core.logging_conf`. This path does not appear to exist. Have you tried `app.logging_conf`?
- I would recommend using an absolute path (with a leading `/`) for the file path in your second example, to ensure the path is correct when running with Docker: `LOGGING_CONF=/app/app/logging_conf.py`.
- In order to use a custom logging class, you will need to tell the logging module to instantiate the class. Instead of `"class": "app.my_logger.MyFormatter"`, try `"()": "app.my_logger.MyFormatter"`.
- The custom logging examples in the _README_ were updated in br3ndonland/inboard@4a280ac. I hope the additional information is helpful.
- The _Dockerfile_ and start script were updated in br3ndonland/inboard#28 and br3ndonland/inboard#29. Give the latest updates a try.
